libraries/lib-wave-track/SequenceSampleCache.cpp

```diff
@@ -0,0 +1,77 @@
+#include "SequenceSampleCache.h"
+
+#include <cassert>
+
+SequenceSampleCache::SequenceSampleCache(
+   SourceSampleGetterFn fn, sampleCount sequenceLength, size_t maxBlockSize)
+    : mGetterFn(std::move(fn))
+    , mMaxBlockSize(maxBlockSize)
+    , mSequenceLength(sequenceLength)
+{
+}
+
+bool SequenceSampleCache::Get(
+   samplePtr buffer, sampleFormat format, sampleCount start, size_t len,
+   bool mayThrow)
+{
+   auto containingBlockIt = std::find_if(
+      mBlocks.begin(), mBlocks.end(),
+      [start](const std::optional<Block>& block) {
+         return block.has_value() && block->Contains(start);
+      });
+   if (containingBlockIt == mBlocks.end())
+   {
+      const auto blockIndex =
+         mLastContainingBlockIndex.has_value() ?
+            (*mLastContainingBlockIndex + 1) % mBlocks.size() :
+            0u;
+      const auto numSamplesToCache =
+         std::min(sampleCount { mMaxBlockSize }, mSequenceLength - start)
+            .as_size_t();
+      auto& block = mBlocks[blockIndex];
+      block.emplace(numSamplesToCache, format, start);
+      if (!mGetterFn(
+             block->buffer.ptr(), format, start, numSamplesToCache, mayThrow))
+      {
+         assert(false);
```

Comment:
I think this misuses the assertion.  If it's part of the contract that the getter function cannot fail, then just give it a void return.  If failure is tolerable (and indeed you handle the case next), then don't assert.

And I would do the latter, knowing what I do about the history.  The `mayThrow` arguments ultimately cause exception propagation when true in case there is some sort of disk failure.  Exceptions lead to a top-level handler in the program that backs out the side effects of a failed operation and restores the last good state out of undo history.  See `AudacityApp::OnExceptionInMainLoop`.

But in some contexts (painting the screen) false is passed around instead to suppress exceptions, and it's expected to fill sample buffers with zero and give that to painting.

Failure conditions because of exogenous device problems or correupted files are contingencies to be handled, and it is not proper to assert that they never happen.

As I said, I don't like new assertions that have no accompanying comments anywhere, unless they are in the middle of the function and very easy to deduce as always correct in the context.

A new assertion should mean "This condition was provable at the time the assertion was written."

You should know, and persuade your reviewer too, that assertion failure is LOGICALLY impossible at the time you add the assertion.

I see assertions as a kind of documentation, not just there for their runtime effects in debug builds.

Now maybe later code modification breaks the proof because it is hard to maintain all the consistency.  That's when the runtime effect of assertions is also valuable.